neqo-transport/src/recovery/sent.rs

```diff
@@ -0,0 +1,382 @@
+    }
+
+    /// Whether the packet should be tracked as in-flight.
+    pub fn cc_in_flight(&self) -> bool {
+        self.ack_eliciting() && self.on_primary_path()
+    }
+
+    /// Declare the packet as lost.  Returns `true` if this is the first time.
+    pub fn declare_lost(&mut self, now: Instant) -> bool {
+        if self.lost() {
+            false
+        } else {
+            self.time_declared_lost = Some(now);
+            true
+        }
+    }
+
+    /// Ask whether this tracked packet has been declared lost for long enough
+    /// that it can be expired and no longer tracked.
+    pub fn expired(&self, now: Instant, expiration_period: Duration) -> bool {
+        self.time_declared_lost
+            .map_or(false, |loss_time| (loss_time + expiration_period) <= now)
+    }
+
+    /// Whether the packet contents were cleared out after a PTO.
+    pub fn pto_fired(&self) -> bool {
+        self.pto
+    }
+
+    /// On PTO, we need to get the recovery tokens so that we can ensure that
+    /// the frames we sent can be sent again in the PTO packet(s).  Do that just once.
+    pub fn pto(&mut self) -> bool {
+        if self.pto || self.lost() {
+            false
+        } else {
+            self.pto = true;
+            true
+        }
+    }
+}
+
+/// A collection for packets that we have sent that haven't been acknowledged.
+#[derive(Debug, Default)]
+pub struct SentPackets {
+    /// The collection.
+    packets: BTreeMap<u64, SentPacket>,
+}
+
+impl SentPackets {
+    pub fn len(&self) -> usize {
+        self.packets.len()
+    }
+
+    pub fn track(&mut self, packet: SentPacket) {
+        self.packets.insert(packet.pn, packet);
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut SentPacket> {
+        self.packets.values_mut()
+    }
+
+    /// Take values from a specified ranges of packet numbers.
+    /// The values returned will be reversed, so that the most recent packet appears first.
+    /// This is because ACK frames arrive with ranges starting from the largest acknowledged
+    /// and we want to match that.
+    pub fn take_ranges<R>(&mut self, acked_ranges: R) -> Vec<SentPacket>
+    where
+        R: IntoIterator<Item = RangeInclusive<PacketNumber>>,
+        R::IntoIter: ExactSizeIterator,
+    {
+        let mut result = Vec::new();
+        // Remove all packets. We will add them back as we don't need them.
+        let mut packets = std::mem::take(&mut self.packets);
+        for range in acked_ranges {
+            // For each acked range, split off the acknowledged part,
+            // then split off the part that hasn't been acknowledged.
+            // This order works better when processing ranges that
+            // have already been processed, which is common.
+            let mut acked = packets.split_off(range.start());
+            let keep = acked.split_off(&(*range.end() + 1));
+            self.packets.extend(keep);
+            result.extend(acked.into_values().rev());
```

Comment:
Most of this file is just moving stuff, but this is a new algorithm.  The old one is https://github.com/mozilla/neqo/blob/bb88aab4525ff74c42b6cc187813c0b0e2813f80/neqo-transport/src/recovery.rs#L313-L336

Reply:
For anyone else wondering why there is no `BTreeMap::take_range`:

https://users.rust-lang.org/t/removing-range-of-elements-from-btreemap/51582